File: node-bindings/classic-node/__test__/ba2_scan_conformance.ts
import {mkdtemp, readFile, rm, writeFile} from "node:fs/promises";
import {tmpdir} from "node:os";
import {join, relative} from "node:path";
import {JsBa2Scanner, scanAllBa2Archives} from "../index.js";

/** Project native archive issue vectors without interpreting the input archive bytes. */
function issues(value: Record<string, any>): Record<string, any> {
    return {dimensions: value.texDims, formats: value.texFrmt, sounds: value.sndFrmt, scripts: value.xseFile};
}

/** Execute default/custom scanner construction, direct reads, discovery and batch convenience. */
export async function observeBa2Scan(fixture: Record<string, any>): Promise<Record<string, any>> {
    if (fixture.operation !== "full") throw new Error("Node uses the full BA2 surface");
    const root = await mkdtemp(join(tmpdir(), "classic-ba2-node-"));
    const path = join(root, "fixture.ba2");
    try {
        await writeFile(path, Buffer.from(fixture.bytes));
        const scanner = new JsBa2Scanner();
        const result = issues(scanner.scanArchive(path));
        const custom = issues(JsBa2Scanner.withXsePatterns([]).scanArchive(path));
        if (JSON.stringify(custom) !== JSON.stringify({
            ...result,
            scripts: []
        })) throw new Error("custom XSE patterns changed unrelated BA2 issue vectors");
        const normalized = (path: string) => relative(root, path).replaceAll("\\", "/");
        return {
            issues: result,
            found: scanner.findBa2Files(root).map(normalized).sort(),
            batch: scanAllBa2Archives(root).map(value => [normalized(value.path), issues(value.issues)]),
            bytes: Array.from(await readFile(path))
        };
    } finally {
        await rm(root, {recursive: true, force: true});
    }
}

File: python-bindings/tests/ba2_scan_conformance.py
"""Read valid, input-only BA2 bytes through native Python scanner APIs."""

import tempfile
from collections.abc import Mapping
from pathlib import Path
from typing import Any


def _issues(value: Any) -> dict[str, list[str]]:
    """Project every actual issue vector and verify public count/query methods."""
    result = {
        "dimensions": value.tex_dims,
        "formats": value.tex_frmt,
        "sounds": value.snd_frmt,
        "scripts": value.xse_file,
    }
    total = sum(len(values) for values in result.values())
    if value.total_count() != total or value.has_issues() is not bool(total):
        raise ValueError("BA2 public summary disagrees with issue vectors")
    return result


def observe_ba2_scan(fixture: Mapping[str, Any]) -> dict[str, Any]:
    """Execute direct, discovery, batch and convenience scans without host archive parsing."""
    import classic_scangame as native

    if fixture["operation"] != "full":
        raise ValueError("Python uses the full scanner surface")
    with tempfile.TemporaryDirectory(prefix="classic-ba2-") as directory:
        root = Path(directory)
        path = root / "fixture.ba2"
        path.write_bytes(bytes(fixture["bytes"]))
        scanner = native.BA2Scanner()
        issues = _issues(scanner.scan_archive(path))
        found = sorted(
            value.relative_to(root).as_posix() for value in scanner.find_ba2_files(root)
        )
        batch = [
            [value.relative_to(root).as_posix(), _issues(result)]
            for value, result in scanner.scan_archives_batch([path])
        ]
        if [
            [value.relative_to(root).as_posix(), _issues(result)]
            for value, result in native.scan_all_ba2_archives(root)
        ] != batch:
            raise ValueError("BA2 directory convenience scan disagrees with batch")
        return {
            "issues": issues,
            "found": found,
            "batch": batch,
            "bytes": list(path.read_bytes()),
        }
